chain/client/src/client_actor.rs

```diff
@@ -1193,6 +1194,25 @@ impl ClientActor {
         // will prioritize processing messages until mailbox is empty. Execution of any other task
         // scheduled with run_later will be delayed.
 
+        // Check block height to trigger expected shutdown
+        if let Ok(head) = self.client.chain.head() {
+            let block = self.client.chain.get_block(&head.last_block_hash).unwrap().clone();
+            let last_final_hash = block.header().last_final_block();
+            let last_final_block_height = self
+                .client
+                .chain
+                .get_block(&last_final_hash)
+                .map_or(0, |block| block.header().height());
+            let block_height_to_shutdown =
+                EXPECTED_SHUTDOWN_AT.load(std::sync::atomic::Ordering::Relaxed);
+            if block_height_to_shutdown > 0 && last_final_block_height >= block_height_to_shutdown {
```

Comment:
do we really need the last 'final' block here?

as we could simply use head.height

(the check triggers is run VERY often - so we have to make sure that we minimise the amount of work we're doing here).